capsules/extra/src/kv_store.rs

```diff
@@ -201,172 +169,229 @@ impl<'a, K: KVSystem<'a, K = T>, T: kv_system::KeyType> KVStore<'a, K, T> {
+                        match op {
+                            Operation::Get | Operation::Set => {
+                                match self.kv.generate_key(unhashed_key, hashed_key) {
+                                    Ok(()) => {
+                                        self.inflight.set(node);
+                                    }
+                                    Err((unhashed_key, hashed_key, e)) => {
+                                        // Issue callback with error.
+                                        self.hashed_key.replace(hashed_key);
+                                        node.operation.clear();
+                                        node.value.take().map(|value| {
+                                            node.client.map(move |cb| {
+                                                cb.get_complete(e, unhashed_key, value);
+                                            });
+                                        });
+                                    }
+                                }
+                            }
+                            Operation::Delete => {
+                                match self.kv.generate_key(unhashed_key, hashed_key) {
+                                    Ok(()) => {
+                                        self.inflight.set(node);
+                                    }
+                                    Err((unhashed_key, hashed_key, e)) => {
+                                        self.hashed_key.replace(hashed_key);
+                                        node.operation.clear();
+                                        node.client.map(move |cb| {
+                                            cb.delete_complete(e, unhashed_key);
+                                        });
+                                    }
+                                }
+                            }
+                        };
+                    });
+                });
+            });
             Ok(())
-        } else {
-            // Another app is already running, queue this app as long as we
-            // don't already have data queued.
-            if self.next_operation.is_none() {
-                self.next_operation.set(Operation::Delete);
-                self.unhashed_key.replace(unhashed_key);
-                self.next_valid_ids.set(perms);
-
-                Ok(())
-            } else {
-                Err((unhashed_key, Err(ErrorCode::BUSY)))
-            }
+        });
+
+        // If we have nothing scheduled, and we have recently done a delete, run
+        // a garbage collect.
+        if ret == Err(ErrorCode::NODEVICE) && self.cleanup.contains(&StateCleanup::CleanupRequested)
+        {
+            self.cleanup.set(StateCleanup::CleanupInProgress);
+            // We have no way to report this error, and even if we could, what
+            // would a user do?
+            let _ = self.kv.garbage_collect();
```

Comment:
Would it make sense to instead have another member that tracks whether garbage collection has failed, and if so errors all further requested operations? That would make it visible to the user that something has broken rather than hiding it.

That will make the driver totally broken if garbage collection fails, but presumably it should never fail, and if it does the driver would likely be totally broken afterwards anyways, right?

Reply:
We sort of have this tradeoff in many Tock capsules, where there are rare underlying errors that could occur, and if they do occur, there is pretty much nothing a user can do to rectify it, and the capsule should probably no longer be used. Some capsules have a 'Error::BROKEN' in their state machine to account for this, but adding such a state comes at a size/complexity cost since you have to check against this possible error state in response to every system call. I am generally in favor of just dropping these errors for now, except in cases where we see people are actually hitting them in practice

Reply:
Is the idea that since this is an extra capsule, it shouldn't have a separate virtualizer?

Reply:
This capsule does two things: 1) add the tock header to the k-v objects, and 2) virtualize the kv_store interface.

We could split those two, but...maybe not right now.